docs/writing_plugins.md

```diff
@@ -0,0 +1,550 @@
+```
+
+`AnyDesc` represents any of the `DescXXX` objects such as `DescFile`, `DescEnum`, `DescMessage`, etc.  The `id` parameter represents the ID of the custom options field definition.
+
+The `scalarType` parameter is the type of the custom option you are searching for.  `ScalarType` is an enum that represents all possible scalar types in the Protobuf grammar
+
+For example, given the following:
+
+```proto
+extend google.protobuf.MessageOptions {
+  optional int32 foo_message_option = 50001;
+}
+extend google.protobuf.FieldOptions {
+  optional string foo_field_option = 50002;
+}
+
+message FooMessage {
+  option (foo_message_option) = 1234;
+
+  int32 foo = 1 [(foo_field_option) = "test"];
+}
+```
+
+The values of these options can be retrieved as follows:
+
+```ts
+const msgVal = findCustomScalarOption(descMessage, 50001, ScalarType.INT32);  // 1234
+
+const fieldVal = findCustomScalarOption(descField, 50002, ScalarType.STRING);  // "test"
+```
+
+#### Message Options
+
+Custom options of a more complex message type can be retrieved via the `findCustomMessageOption` function.  It returns a concrete type with fields populated corresponding to the values set in the proto file.
+
+```ts
+export function findCustomMessageOption<T extends Message<T>>(
+  desc: AnyDesc,
+  id: number,
+  msgType: MessageType<T>
+): T | undefined {
+```
+
+The `msgType` parameter represents the type of the message you are searching for.  Note that this can be a message generated from a proto file or can be a type created at runtime via `makeMessageType`.
+
+For example, given the following proto files:
+
+```proto
+// custom_options.proto
+
+extend google.protobuf.MethodOptions {
+  optional ServiceOptions service_method_option = 50007;
+}
+
+message ServiceOptions {
+  int32 foo = 1;
+  string bar = 2;
+  oneof qux {
+    string quux = 3;
+  }
+  repeated string many = 4;
+  map<string, string> mapping = 5;
+}
+```
+
+```proto
+// service.proto
+
+import "custom_options.proto";
+
+service FooService {
+  rpc Get(GetRequest) returns (GetResponse) {
+    option (service_method_option) = { 
+        foo: 567, 
+        bar: "Some string", 
+        quux: "Oneof string",
+        many: ["a", "b", "c"],
+        mapping: [{key: "testKey", value: "testVal"}]
+    };
+  }
+}
+```
+
+You can retrieve the options using a generated type by first generating the file which defines the custom option type.  Then, import and pass this type to the `findCustomMessageOption` function.
+
+```ts
+import { ServiceOptions } from "./gen/proto/custom_options_pb.js";
+
+const option = findCustomMessageOption(method, 50007, ServiceOptions)
+
+console.log(option);
+/*
+ * {
+ *     foo: 567, 
+ *     bar: "Some string", 
+ *     quux: "Oneof string",
+ *     many: ["a", "b", "c"],
+ *     mapping: [{key: "testKey", value: "testVal"}]
+ * }
+ */
+ ```
+ 
+ Alternatively, you can use a type created at runtime:
+
+```ts
+const opts = proto3.makeMessageType("ServiceOptions", [
```

Comment:
Same here. 

There _are_ use cases for creating a type at run time, but not hard coded like this, but via `createRegistryFromDescriptors()`. This would allow you to parse custom options and - for example - turn them into JSON without knowing their concrete type. Not sure we need to go into those details in this doc?